oxo.py

```diff
@@ -1,26 +1,195 @@
 class OxoBoard:
-    def __init__(self):
-        """ The initialiser. Initialise any fields you need here. """
-        raise NotImplementedError("TODO: implement __init__")
+
+    PLAYER_y = 0
+    PLAYER_X = 1
+
+    def __init__(self, grid_width=3, grid_height=3, win_row_amount=3):
+        """ The initialiser. Initialise any fields you need here.
+
+        :param grid_width:     width of board
+        :param grid_height:    height of board
+        :param win_row_amount:
+        """
+        self.grid_size = {"width": grid_width, "height": grid_height}
+        self.total_cells = grid_width * grid_height
+        self.grid = {}
+        self.win_row_amount = win_row_amount
 
     def get_square(self, x, y):
         """ Return 0, 1 or 2 depending on the contents of the specified square. """
-        raise NotImplementedError("TODO: implement get_square")
+        if (x, y) in self.grid:
+            return self.grid[(x, y)]
+        else:
+            return 0
 
     def set_square(self, x, y, mark):
         """ If the specified square is currently empty (0), fill it with mark and return True.
             If the square is not empty, leave it as-is and return False. """
-        raise NotImplementedError("TODO: implement set_square")
+        if (x, y) not in self.grid:
+            self.grid[(x, y)] = mark
+            return True
+        else:
+            return False
 
     def is_board_full(self):
         """ If there are still empty squares on the board, return False.
             If there are no empty squares, return True. """
-        raise NotImplementedError("TODO: implement is_board_full")
+        return len(self.grid) == self.total_cells
 
     def get_winner(self):
         """ If a player has three in a row, return 1 or 2 depending on which player.
             Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+        for player_index in range(1, 3, 1):
+            # Horizontal
+            winner = self.check_hoz(player_index)
+            if winner != 0:
+                return winner
+            # Vertical
+            winner = self.check_vert(player_index)
+            if winner != 0:
+                return winner
+            # Diagonals
+            winner = self.check_diagonals(player_index)
+            if winner != 0:
+                return winner
+        return 0
+
+    def check_hoz(self, player):
```

Comment:
`check_hoz` and `check_vert` are quite similar -- could they be combined?

Reply:
On a reflection of this I could of added 'X_size' and 'Y_size' to the function pramas and flip the values when called for instance 
check_row(self.grid_size["width"], self.grid_size["height"], player) for horizontal and check_row(self.grid_size["height"], self.grid_size["width"], player) for vertical

I would also have to change some of the variables in the function to the pramas as well but it would be pretty straight forward